Optimize Image Stuff By Evaluating Plugs in Global Context

danieldresser opened the pull request:
This is just being put up for comment - my guess is that some of these commits may be something that you think are a good idea anyway, some of them you probably don't want, and some of them you might only want if we can come up with good tests to prove performance.

I'm still testing to what extent I actually need these changes, so you could just ignore this until I've got some better data.

It's really hard to evaluate performance of this stuff in isolation. I was seeing significant costs to evaluating channelNames and dataWindow in a test, but I never saw much cost for format - but this depends entirely on the upstream network - it's tough to say whether you possibly build an upstream network that would make it costly to evaluate format plugs.

For a crazy test that I was benchmarking on, this originally sped up hashing by a factor of 2, and overall runtime by up to 20% - but after doing this work, I took a look at what was left, and cleaned up some of the stuff I was doing ( avoiding OSLImage seems to have been the big one ), and now this stuff seems to matter a lot less.  Which makes some sense, this stuff is only expensive if you have something upstream that blows up the hashing time, but now I'm not sure how useful this is.

I will evaluate further once I've got a full setup together of my current image processing thing.

johnhaddon commented:
Thanks Daniel - looks pretty good at first glance. I'll take a closer look when you have your latest results available and we have a better idea of what works and what doesn't. As you say though, it's very dependent on the input graph, and it might be that we're slightly hurting the performance for the simple cases while improving the performance for the complex cases.

One thing I might take a look at is making a new `TemporaryContext` class as a base class for your global context. This would just hide away the `new Context( Borrowed ); Context::Scope` pair that we're repeating everywhere, and give us the possibility to further improve performance later by avoiding the heap allocation for small contexts. I've been meaning to do something like this for a while, so it seems to make sense now if we're going to expose the GlobalContext publicly.

danieldresser commented:
Given the dependence on the input graph, I'm a bit unsure how much profiling you want me to go into here.  On my current heavy network in practice, running on the farm, batched to 5 frames at a time to benefit from caching of things that are TimeWarped, I'm seeing:
Before:  Avg: 350 seconds, Max: 453 seconds ( per 5 frames )
After:  Avg: 204 seconds, Max: 281 seconds ( per 5 frames )

I wasn't controlling for farm heterogenity, so the percentages won't be completely accurate, but this approximately matches what I'm seeing locally.  There is also quite a perceptible difference in feel interactively when clicking between expensive to hash things that you've already computed.

I'm not sure what the best way to quantify your fears about downsides in simple cases is - hopefully in simple cases, we're not seeing huge amounts of hashing time?

I wonder if it would be worth you taking a look through these PRs, and sort them into things you think are likely a good idea ( after your TemporaryContext addition ), and things that feel unnecessary to you?  Some of these seem like very probably a generally positive thing to me.  If there are some that feel unnecessary to you, I could try running my big production test without them, and if they turn out to be useful in the case I need I could target those areas a bit more specifically in trying to construct a performance test that isn't quite as complicated?

danieldresser commented:
Oh, also, I wonder if I should break out the one commit to performanceMonitor into a separate PR - it just adds an output for total hashing time, which is pretty useful when looking at this stuff.

johnhaddon commented on src/GafferImage/Resample.cpp:
```diff
@@ -454,7 +454,10 @@ void Resample::hashChannelData( const GafferImage::ImagePlug *parent, const Gaff
 	V2f ratio, offset;
-	ratioAndOffset( matrixPlug()->getValue(), ratio, offset );
+	{
+        ImagePlug::GlobalContext c( context );
```
Indentation.

johnhaddon commented on src/GafferImage/Resample.cpp:
```diff
@@ -494,7 +497,10 @@ void Resample::hashChannelData( const GafferImage::ImagePlug *parent, const Gaff
 	V2f ratio, offset;
-	ratioAndOffset( matrixPlug()->getValue(), ratio, offset );
+	{
+        ImagePlug::GlobalContext c( context );
```
Indentation.

johnhaddon commented on python/GafferImageTest/ImageTransformTest.py:
```diff
@@ -171,10 +171,10 @@ def testDisabled( self ) :
 			t["transform"]["rotate"].setValue( 90 )
 			t["enabled"].setValue( True )
-			self.assertNotEqual( r["out"].hash(), t["out"].hash() )
+			#self.assertNotEqual( r["out"].hash(), t["out"].hash() )
```
Can you explain this?

danieldresser replied:
This was part of my commit "DONT MERGE : For illustration, this is how I was locating code that evaluated global plugs in a non-global context", which added asserts to ensure that global plugs were forced to never be called in non-global contexts.  By explicitly setting the tile origin, and then hashing the image plug ( which implicitly hashes all it's children including the global plugs ), this test violates that rule.

I was kind of guessing that you wouldn't want those asserts anyways, I was thinking that this would be an optimization we would do only when it was profiled to matter.  But if you're mostly OK with this batch of commits, maybe we should think about keeping these asserts in there?

Failing to use global scope for global plugs is a performance hazard which may not show up in most testing but has the potential to dramatically slow down users with obscure and complex setups.

As for this test, do you think there's a good reason why it explicitly sets tileOrigin and channel and calls ImagePlug.hash() instead of calling ImagePlug.imageHash()?  And actually, what's it doing with CapturingSlot that it doesn't appear to actually use?

danieldresser replied:
Oh, you were going to get back to me on whether it would be reasonable to use imageHash() here?

johnhaddon replied:
> Oh, you were going to get back to me on whether it would be reasonable to use imageHash() here?

Yeah, that test is pretty bogus isn't it? We should remove the context and the CapturingSlot and just call `imageHash()` instead.

johnhaddon commented on src/GafferImage/Merge.cpp:
```diff
@@ -199,7 +199,7 @@ void Merge::hashChannelData( const GafferImage::ImagePlug *output, const Gaffer:
 		}
 
-		IECore::ConstStringVectorDataPtr channelNamesData = (*it)->channelNamesPlug()->getValue();
+		IECore::ConstStringVectorDataPtr channelNamesData = (*it)->channelNames();
```
It seems unfortunate that the context construction is inside the loop here. Can we at least share a single GlobalContext with the dataWindow computation below?

johnhaddon commented on src/GafferImage/Merge.cpp:
```diff
@@ -278,7 +278,7 @@ IECore::ConstFloatVectorDataPtr Merge::merge( F f, const std::string &channelNam
 		}
 
-		IECore::ConstStringVectorDataPtr channelNamesData = (*it)->channelNamesPlug()->getValue();
+		IECore::ConstStringVectorDataPtr channelNamesData = (*it)->channelNames();
```
Same here - can we share a GlobalContext with the dataWindow access below? Just to be clear, not by adding a getChannelNamesAndDataWindow accessor on ImagePlug, just by manually making a GlobalContext here.

johnhaddon commented on src/GafferImage/VectorWarp.cpp:
```diff
@@ -184,9 +184,9 @@ void VectorWarp::hashEngine( const std::string &channelName, const Imath::V2i &t
 
 	h.append( tileOrigin );
-	vectorPlug()->dataWindowPlug()->hash( h );
+	h.append( vectorPlug()->dataWindowHash() );
```
Looks like this is another one where we could make one GlobalContext if we did that ourselves?

johnhaddon commented on include/GafferImage/ImagePlug.h:
```diff
@@ -135,6 +136,39 @@ class ImagePlug : public Gaffer::ValuePlug
 		//@}
 
+		/// @name Other convenience accessors
+		/// For things that should never depend on channelName and tileOrigin
```
We should document the reason for these accessors, providing the guideline that they should be used when accessing from a per-tile compute/hash, but not from a compute/hash that is already in a global context.

johnhaddon commented on src/GafferImage/CDL.cpp:
```diff
@@ -131,6 +131,8 @@ bool CDL::affectsTransform( const Gaffer::Plug *input ) const
 
 void CDL::hashTransform( const Gaffer::Context *context, IECore::MurmurHash &h ) const
 {
+	ImagePlug::GlobalContext c( context );
```
Can the responsibility for this be pushed up onto the OpenColorIOTransform class? That seems preferable to having to remember it in all the base classes...

johnhaddon commented on include/Gaffer/PerformanceMonitor.h:
```diff
@@ -117,6 +119,7 @@ class PerformanceMonitor : public Monitor
 		// Then when we want to query it, we collate it into m_statistics.
 		void collate() const;
 		mutable StatisticsMap m_statistics;
+		mutable Statistics m_combinedStatistics;
```
Can you add a commit comment mentioning the binary compatibility break?

johnhaddon commented on src/Gaffer/MonitorAlgo.cpp:
```diff
@@ -318,6 +319,9 @@ struct FormatStatistics
 
 		std::stringstream s;
+
+		s << "Total " << metric.description() << " : " <<  metric( combinedStatistics ) << "\n";
```
I'm finding this formatting really awkward to read. Could we put all the totals in a single section on their own, with the same indentation we have for the other sections?

```
PerformanceMonitor Summary :

  Time spent in hash and compute processes : x.xx
  Time spent in hash processes             : x.xx
  Time spent in compute processes          : x.xx
  ...
```

danieldresser commented:
I thought about this for a while, and felt a bit uncertain about the best way to clean it up, and talked to Don a bit about it, and then got distracted and did my last two main optimization goals from earlier:
* driverChannel plug on Median ( PR coming as soon as I figure out a good way to write the unit test.  I've got a nifty little Gaffer network that tests it well, but it would be a bit complicated to explain in code )
* optimizations to Warp ( speculative PR up )

The conclusions from my conversation with Don ( well, I shouldn't blame him for this, my conclusions, with some input from him ):
* GlobalContext is a hilariously vague name.  Also it doesn't indicate that it acts as a Scope.  How about I propose something completely different like WholeImageScope, and see if that suggests anything reasonable to you.
* if the accessors are actually things that you should only use if you're not already in a global context, then the names of them should probably reflect that - dataWindowHashInImageScope or something ugly like that?
* if the calling code needs to be aware of the details of contexts anyways, then it kinda feels like these accessors shouldn't exist.  The only real reason for them is to save typing all over the place?

Don suggested an approach like a generic wrapper function:
```
template<typename PlugType>
typename PlugType::ValueType imageScopeValue(const PlugType& plug)
{
   const Gaffer::Context *c = Gaffer::Context::current();
   Gaffer::Context n (*c, Gaffer::Context::Copied );

   Gaffer::Context::Scope s(&n);

   n.remove("a");
   n.remove("b");

   return plug.getValue();
}
```
called like:
```
imageScopeValue( *dataWindowPlug() );
```

I also thought about a templated wrapper version of getValue that I guess would live right down in Gaffer::Plug, and would instantiate a templated class before calling the standard getValue, so you could just do:
```
dataWindowPlug()->getValue< GlobalContext >();
```

I don't really want to make my life getting this merged more difficult, but it also feels a bit like I'm introducing some weird ugliness that we're going to regret.  If you're happy with it though, I'll go ahead and clean it up as is ( though I do feel like I should put clearer names on those accessors ).

Oh, and one other random thought: if creating contexts is expensive, should GlobalContext check if either of the variables it's affecting exist already, and not create a new context if they don't?  I guess it really should be the client's responsibility not to create a GlobalContext when already in a GlobalContext.

johnhaddon commented:
> * GlobalContext is a hilariously vague name.  Also it doesn't indicate that it acts as a Scope.  How about I propose something completely different like WholeImageScope, and see if that suggests anything reasonable to you.

I think my struggling with the naming before was one of the reasons this stayed private for so long. But I don't think the name is vague - you've taken it out of context (excuse the pun) above, and it's actually `ImagePlug::GlobalContext`. That makes it clear that it's about images, and about the global aspects of them. WholeImage makes me think it would relate to all the tiles somehow.

I buy the thing about making it clearer that it's a Scope though. Let me explain some other stuff that needs to happen with the Context class at some point, and see if it gets us anywhere with naming....

We construct a lot of temporary contexts in our hash/compute implementations, and the copying of all the variables within the context was a big overhead. So I introduced the concept of ownership, so you could construct a non-owning copy of a context, promising that it would exist only as long as the thing being copied, and could just reference the data directly without ownership. So this is the current idiom :

```
ContextPtr temporary = new Context( *context, Context::Borrowed ); // Promise c won't outlive context
temporary->set( "a", 10 );
Context::Scope scope( temporary );
...
```

But this is _aweful_. You have to allocate the context, and you get a smart pointer to it - just the thing that would facilitate it outliving the source. You also have to read the docs for the Context::Borrowed argument to get what it's all about. And you have to do the scoping in a separate step. Rubbish.

So I think this should become something more like this, which is basically the same as GlobalContext in terms of implementation, but also allows subsequent set/remove :

```
TemporaryContext c( *context );
c.remove( "b" );
c.set( "a", 10 );
...
```

This enforces the lifetime because it's on the stack, and also gives us further opportunities to optimise behind the scenes later (we could potentially use the stack rather than allocate a new context internally, but this has a whole bunch of issues I won't get into here).

One of the problems with this is how TemporaryContext relates to Context. It sounds a bit like it should be a derived class, but we can't implement it that way right now without lots of upheaval. And maybe that wouldn't make make sense anyway because it does it's own scoping?

Now I've talked my way through it, I think the name `Context::EditableScope` is quite nice. What do you think?

If that sounds OK, then perhaps `ImagePlug::GlobalScope`? That doesn't make it super clear that it's related to contexts, but when seen with the constructor arguments it might make sense?

```
ImagePlug::GlobalScope scope( *context );
...
```

> * if the accessors are actually things that you should only use if you're not already in a global context, then the names of them should probably reflect that - dataWindowHashInImageScope or something ugly like that?

I'm not sure I can bear that level of verbosity - we already have equivalent non-verbose accessors for ScenePlug and it's felt OK, so if we have accessors, let's stick with the names you've already got.

> * if the calling code needs to be aware of the details of contexts anyways, then it kinda feels like these accessors shouldn't exist.  The only real reason for them is to save typing all over the place?

I think the calling code _does_ need to be aware of the details of the context, so yes, the only benefit is less typing. It also introduces the pitfalls that I already commented on in one of your commits - it hides an expensive context construction, so you can easily inadvertently make a whole bunch of temporary contexts in a loop because you're using a couple of different accessors.

> Don suggested an approach like a generic wrapper function:
>```
>template<typename PlugType>
>typename PlugType::ValueType imageScopeValue(const PlugType& plug)
>{
>
>    const Gaffer::Context *c = Gaffer::Context::current();
>    Gaffer::Context n (*c, Gaffer::Context::Copied );
> 
>    Gaffer::Context::Scope s(&n);
> 
>    n.remove("a");
>    n.remove("b");
> 
>    return plug.getValue();
> }
> ```
> called like:
> ```
> imageScopeValue( *dataWindowPlug() );
> ```

Although this does make it clear that you're getting a value in a particular scope, it suffers from the exact same problem as the accessors - getting multiple values or using it in a loop incurs unnecessary overhead.

> I also thought about a templated wrapper version of getValue that I guess would live right down in >Gaffer::Plug, and would instantiate a templated class before calling the standard getValue, so you >could just do:
>```
>dataWindowPlug()->getValue< GlobalContext >();
>```

This has the same problem. While we're thinking of solving the issue with language features (templating in this case), I did see an [interesting construct using C++11 lambdas](https://herbsutter.com/elements-of-modern-c-style/) the other day (scroll down to the Lambda Functions and Algorithms section). If I've got the gist, we'd define a template function which takes a scope and a lambda, and that would give us some nice syntactic sugar. Mirroring python, we might call the function `with`, and it'd look something like this :

```
with( Context::EditableScope( context ), [&]{
     ...
})
```

Actually, I'm not sure how we'd pass the editable scope into the lambda - @donboie, I expect you might know? I'm also not sure the little bit of syntactic sugar this buys is all that much of an improvement, and we're still not in 2011 yet anyway.

> I don't really want to make my life getting this merged more difficult, but it also feels a bit like I'm introducing some weird ugliness that we're going to regret.  If you're happy with it though, I'll go ahead and clean it up as is ( though I do feel like I should put clearer names on those accessors ).

Yeah, me neither - I thought we were pretty much there with this already. I've enjoyed the opportunity to talk through some of the details of the context scoping though, and think I'm fairly happy with this proposal :

- I make a Context::EditableScope class as described above, and remove all the heinous `new Context( Borrowed )` usage (although it'll be available for backwards compatibility for a while still).
- You derive GlobalContext from Context::EditableScope, and call it `ImagePlug::GlobalScope`.
- We either keep the accessors as they currently are (simple names matching the ScenePlug equivalent) or we remove them on the grounds that they're a performance hazard.

How does that sound?

> Oh, and one other random thought: if creating contexts is expensive, should GlobalContext check if either of the variables it's affecting exist already, and not create a new context if they don't?  I guess it really should be the client's responsibility not to create a GlobalContext when already in a GlobalContext.

Yeah, I think it's the client's responsibility. If we want optimisation we might be able to push it all the way down into `EditableContext::remove()` anyway, but let's leave that all for later.

danieldresser commented:
> * I make a Context::EditableScope class as described above, and remove all the heinous new Context( Borrowed ) usage (although it'll be available for backwards compatibility for a while still).
> * You derive GlobalContext from Context::EditableScope, and call it ImagePlug::GlobalScope.
> * We either keep the accessors as they currently are (simple names matching the ScenePlug equivalent) or we remove them on the grounds that they're a performance hazard.
> How does that sound?

Yeah, sounds fine.  EditableScope sounds good for making the places where I need to do something special a bit nicer ( like removing just channelName without removing tileOrigin ).  Still feels a bit weird to me to have a simple name but a complex set of reasoning for when you should call it, but point taken that the scenePlug stuff is the same.

Is EditableScope something you're going to be able to bang out quickly?

johnhaddon commented:
> Is EditableScope something you're going to be able to bang out quickly?

I started noodling around with it this afternoon actually, to get a feel for what it was like in practice. It's definitely an improvement I think - less code at the client end, less opportunity to use it wrong, and in a form we could optimise more later. The only thing that felt a bit odd was that I was constructing EditableScopes but always calling them `tmpContext` or `instanceContext` or something else contexty. But then again, I can't think of a better name, so unless you can let's go with it...

I should be able to get this tidied up for review tomorrow - is that OK?

danieldresser commented:
Sounds good!

johnhaddon commented:
OK, I've got the Context improvements up for review in #2034. Perhaps you could review them Daniel, and then rebase your work on top once you're happy and have merged? As well as adding ScenePlug::GlobalScope, I also added ScenePlug::PathScope and ScenePlug::SetScope, which are useful in their own right but I think also help to make it more obvious what GlobalScope is.

I've updated everything except GafferImage/OSLImage to avoid Context::Borrowed and use Scopes instead. I left  GafferImage alone because I didn't want to introduce clashes with your work. If you could make the equivalent changes in GafferImage that would be brilliant - do you think that's doable?

danieldresser commented:
> If you could make the equivalent changes in GafferImage that would be brilliant - do you think that's doable?

Yeah, I should be able to take a look at that.  Would be nice to get the WarpPerformance PR merged first, because I would make a change on top of that here - could you take a look at that?

johnhaddon commented:
Hey Daniel,
Sorry I didn't get to that one yet. I'd like to finish on time tonight though so perhaps someone else could take a quick look? Have a good weekend...
Cheers...
John

danieldresser commented:
OK, I've now addressed your code comments, rebased on the new EditableScope branch and updated everything.  The last 3 commits are just related to using EditableScope everywhere, and could be a separate PR from this.

I ended up removing the accessors functions - if we're saying that it is a performance related error to ever call two accessor functions from the same scope, and you should instead use a GlobalScope and getValue, then having the accessors feel like too much of a trap ( and there weren't a huge number of uses left once I removed the from all the places where I could reuse a scope ).

Reusing one scope on each iteration of Merge instead of making two made a noticeable difference in our worst case simple test of merging 10 Constants - the difference from the previous code is now statistically insignificant without running very large batches of tests.  I assume that the context construction is still slowing things down a bit, but the hashing reduction probably also has some impact, and they are cancelling out.

Still haven't run a final confirmation test on a production compositing network, but aside from that this is hopefully just about good to merge.

johnhaddon commented on src/GafferImage/ImageSwitch.cpp:
```diff
@@ -44,24 +44,15 @@ using namespace GafferImage;
-/// \todo If we introduce a TemporaryContext base class in Context.h
-/// then this should derive from that.
-struct ImageSwitchIndexContext
+struct ImageSwitchIndexContext : public Gaffer::Context::EditableScope
```
I think we can ditch this class entirely now can't we, and just use `ImagePlug::GlobalScope` directly in SwitchTraits?

johnhaddon commented:
I feel a little bit bad suggesting this, but not _too_ bad since I just did two equivalent passes over GafferScene myself - what do you think to adding an `ImagePlug::ChannelDataScope` class, with separate `setTileOrigin()` and `setChannelName()` methods? I feel like it might make a worthwhile improvement in many of those Borrowed->EditableScope changes in the final commit.

johnhaddon commented on include/Gaffer/MonitorAlgo.h:
```diff
@@ -65,6 +65,7 @@ enum PerformanceMetric
 
 std::string formatStatistics( const PerformanceMonitor &monitor, size_t maxLinesPerMetric = 50 );
 std::string formatStatistics( const PerformanceMonitor &monitor, PerformanceMetric metric, size_t maxLines = 50 );
+std::string formatTotalStatistics( const PerformanceMonitor &monitor );
```
Any reason this is a separate function rather than just being functionality of `formatStatistics()`? I can't think of a reason to not want the summary, and if I wanted only the summary, perhaps I could pass `maxLinesPerMetric = 0`. But if you do think it's important to keep it, then perhaps calling it `formatCombinedStatistics()` would make it more obvious that it outputs `monitor.combinedStatistics()`?

danieldresser commented:
Comments addressed, and should be ready for merging now that the context API changes are in.

johnhaddon commented on include/GafferImage/ImagePlug.h:
```diff
@@ -135,6 +136,26 @@ class ImagePlug : public Gaffer::ValuePlug
+			GlobalScope( const Gaffer::Context *context );
+		};
+
+		/// Utility class to scope a temporary copy of a context,
```
This comment seems to apply to the GlobalScope rather than the ChannelDataScope.

johnhaddon commented:
Looks good Daniel - could you just fix up the comment on the ChannelDataScope and then merge?
Cheers...
John

danieldresser commented:
Cool, will do!  I forgot to mention - I did test this on a production slap comp, and it sped up total time by 15%, and halved hashing time.  So hopefully everyone will see at least a little bit of improvement from this
